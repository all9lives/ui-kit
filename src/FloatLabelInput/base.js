import React, { PureComponent } from 'react'
import { css } from 'styled-components'
import defaultProps from './defaultProps'
import Color from '../utils/Color'

const base = css`
  position: relative;
  width: 100%;
`

export const labelBase = css`
  color: ${props => props.isFocus
    ? (props.labelFocusColor || Color.primary)
    : (props.labelColor || Color.black)};
`

export const inputBase = css`
  position: relative;
`

export const getComponent = ({ Container, Label, Input }) => class Component extends PureComponent {
  static defaultProps = defaultProps;
  state = {
    isFocus: false
  }
  handleFocus = (e) => {
    if (this.props.onFocus) {
      this.props.onFocus(e)
    }
    this.setState({ isFocus: true })
  }
  handleBlur = (e) => {
    if (this.props.onBlur) {
      this.props.onBlur(e)
    }
    this.setState({ isFocus: false })
  }
  render () {
    const { isFocus } = this.state
    const { value, placeholder, labelColor, labelFocusColor } = this.props
    return (
      <Container>
        <Label
          size='small'
          show={!!value}
          isFocus={isFocus}
          labelColor={labelColor}
          labelFocusColor={labelFocusColor}
        >
          {placeholder}
        </Label>
        <Input
          {...this.props}
          onFocus={this.handleFocus}
          onBlur={this.handleBlur}
        />
      </Container>
    )
  }
}

export default base
