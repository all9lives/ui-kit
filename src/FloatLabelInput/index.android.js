import React, { PureComponent } from 'react'
import styled from 'styled-components/native'
import { Animated } from 'react-native'
import base, { getComponent, labelBase, inputBase } from './base'
import View from '../View'
import BaseInput from '../Input'
import Text from '../Text'

export const Container = styled(View)`
  ${base}
`

const Label = styled(Text)`
  ${labelBase}
`

const Input = styled(Animated.createAnimatedComponent(BaseInput))`
  ${inputBase}
`

export class AnimatedLabel extends PureComponent {
  state = {
    value: new Animated.Value(0)
  }
  componentWillReceiveProps (nextProps) {
    if (nextProps.show !== this.props.show) {
      Animated.timing(this.state.value, {
        toValue: nextProps.show ? 1 : 0,
        duration: 300
      }).start()
    }
  }
  render () {
    const { value } = this.state
    const position = {
      position: 'absolute',
      top: 2,
      left: 8
    }
    const translateY = value.interpolate({
      inputRange: [0, 1],
      outputRange: [8, 0]
    })
    return (
      <Animated.View style={{ opacity: value, transform: [{ translateY }], ...position }}>
        <Label {...this.props} />
      </Animated.View>
    )
  }
}

export class AnimatedInput extends PureComponent {
  state = {
    value: new Animated.Value(0)
  }
  componentWillReceiveProps (nextProps) {
    if (nextProps.value !== this.props.value) {
      Animated.timing(this.state.value, {
        toValue: nextProps.value ? 1 : 0,
        duration: 300
      }).start()
    }
  }
  render () {
    const { value } = this.state
    const paddingTop = value.interpolate({
      inputRange: [0, 1],
      outputRange: [10, 18]
    })
    const paddingBottom = value.interpolate({
      inputRange: [0, 1],
      outputRange: [10, 2]
    })
    return (
      <Input style={{ paddingTop, paddingBottom }} {...this.props} />
    )
  }
}

export default getComponent({ Container, Label: AnimatedLabel, Input: AnimatedInput })
