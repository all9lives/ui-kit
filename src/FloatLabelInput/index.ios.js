import FloatLabelInput, { AnimatedInput, AnimatedLabel } from './index.android'

export { AnimatedInput }
export { AnimatedLabel }
export default FloatLabelInput
