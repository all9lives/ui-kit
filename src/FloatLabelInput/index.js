import styled from 'styled-components'
import base, { getComponent, labelBase, inputBase } from './base'
import View from '../View'
import BaseInput from '../Input'
import Text from '../Text'
import Unit from '../utils/Unit'

const getOpacity = (props) => {
  if (props.show) {
    return props.isFocus ? 1 : 0.5
  }
  return 0
}

const Container = styled(View)`
  ${base}
  width: 100%;
`

export const Label = styled(Text)`
  ${labelBase}
  position: absolute;
  z-index: 1;
  top: ${Unit.parse(2)};
  left: ${Unit.parse(8)};
  opacity: ${getOpacity};
  transform: ${props => props.show ? 'translateY(0)' : `translateY(${Unit.parse(8)})`};
  pointer-events: none;
  transition: 300ms ease-in;
  transition-property: transform, opacity, color;
`

export const Input = styled(BaseInput)`
  ${inputBase}
  padding-top: ${props => props.value ? Unit.parse(18) : Unit.parse(10)};
  padding-bottom: ${props => props.value ? Unit.parse(2) : Unit.parse(10)};
  transition: 300ms ease-in;
  transition-property: padding-top, padding-bottom;
`

export default getComponent({ Container, Label, Input })
