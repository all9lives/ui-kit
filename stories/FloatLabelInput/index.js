import React from 'react'
import createStory from '../../storybook/createStory'
import InputWrapper from '../../storybook/createStory/InputWrapper'
import FloatLabelInput from '../../src/FloatLabelInput'

createStory('FloatLabelInput', module)
  .add('default', () => (
    <InputWrapper>
      <FloatLabelInput />
    </InputWrapper>
  ))
