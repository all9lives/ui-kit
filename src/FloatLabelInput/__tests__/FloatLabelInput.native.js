/* eslint-env jest */
import React from 'react'
import { shallow } from 'enzyme'
import FloatLabelInput from '../index.ios.js'

describe('FloatLabelInput.native', () => {
  it('renders a snapshot', () => {
    const wrapper = shallow(
      <FloatLabelInput />
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('renders a snapshot with `handleFocus`', () => {
    const wrapper = shallow(
      <FloatLabelInput />
    )
    wrapper.instance().handleFocus()
    wrapper.update()
    expect(wrapper).toMatchSnapshot()
  })

  it('renders a snapshot with `handleBlur`', () => {
    const wrapper = shallow(
      <FloatLabelInput />
    )
    wrapper.instance().handleFocus()
    wrapper.update()
    wrapper.instance().handleBlur()
    wrapper.update()
    expect(wrapper).toMatchSnapshot()
  })
})
