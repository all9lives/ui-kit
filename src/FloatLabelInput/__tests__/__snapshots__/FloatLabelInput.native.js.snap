// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FloatLabelInput.native renders a snapshot 1`] = `
<Styled(Styled(View))>
  <AnimatedLabel
    isFocus={false}
    show={false}
    size="small"
  >
    Placeholder
  </AnimatedLabel>
  <AnimatedInput
    onBlur={[Function]}
    onFocus={[Function]}
    placeholder="Placeholder"
  />
</Styled(Styled(View))>
`;

exports[`FloatLabelInput.native renders a snapshot with \`handleBlur\` 1`] = `
<Styled(Styled(View))>
  <AnimatedLabel
    isFocus={false}
    show={false}
    size="small"
  >
    Placeholder
  </AnimatedLabel>
  <AnimatedInput
    onBlur={[Function]}
    onFocus={[Function]}
    placeholder="Placeholder"
  />
</Styled(Styled(View))>
`;

exports[`FloatLabelInput.native renders a snapshot with \`handleFocus\` 1`] = `
<Styled(Styled(View))>
  <AnimatedLabel
    isFocus={true}
    show={false}
    size="small"
  >
    Placeholder
  </AnimatedLabel>
  <AnimatedInput
    onBlur={[Function]}
    onFocus={[Function]}
    placeholder="Placeholder"
  />
</Styled(Styled(View))>
`;
